rgw: allow to set ssl options and ciphers for beast frontend

trociny opened the pull request:
Two new conf keys are added for "beast" framework:

- ssl_options: a colon separated list of ssl context options,
  documented in boost's ssl::context_base;

- ssl_ciphers: a colon separated list of ciphers, documented
  in openssl's ciphers(1) manual.

Example:

  rgw frontends = beast ...  ssl_options=default_workarounds:no_tlsv1:no_tlsv1_1 ssl_ciphers=HIGH:!aNULL:!MD5

Fixes: https://tracker.ceph.com/issues/50932
Signed-off-by: Mykola Golub <mgolub@suse.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

trociny commented:
The options are documented in [1] and cipher configuration is documented in [2,3].

I am going to update the rgw documentation when (if) there is agreement about the approach.

[1] https://www.boost.org/doc/libs/1_65_1/doc/html/boost_asio/reference/ssl__context_base.html
[2] https://linux.die.net/man/3/ssl_set_cipher_list
[3] https://linux.die.net/man/1/ciphers

k0ste commented:
@trociny maybe are you aware, do a beast have a option for 302 redirects on 80 port (http -> https) civetweb like `ports=443+80r`?

yehudasa reviewed (commented):
lgtm, need to update docs

trociny commented:
> @trociny maybe are you aware, do a beast have a option for 302 redirects on 80 port (http -> https) civetweb like `port=0.0.0.0:80r+443s`?

Sorry, I have no idea about this.

trociny commented:
The doc is updated.

mdw-at-linuxbox commented:
Other than the documentation nit I noted, I don't see anything bad here.  I'd be tempted to create a constant map for the ssl options, but I'm not convinced totally convinced that's better in C++.  302 redirects could be a separate commit.

trociny commented:
> Other than the documentation nit I noted, I don't see anything bad here.

Sorry, what nit do you mean? I don't see any comment about it.

> I'd be tempted to create a constant map for the ssl options, but I'm not convinced totally convinced that's better in C++.

If it had been used in more than one place sure I would have made a map. But for the current code I don't see much benefit in it. And it can be easily changed in future if we need. (E.g. I was thinking we could add options validator to common/options.cc, but if we really wanted it I think we would add it not only for ssl options but for all other backend options, so it looks like a separate mini-project).

k0ste commented:
> If it had been used in more than one place sure I would have made a map. But for the current code I don't see much benefit in it.

Agreed, considering that next release - where beast is one and the only frontend

mdw-at-linuxbox commented:
Line 69 in doc/radosgw/frontends.rst says column; I suspect you want to say colon.  [ I see my comment above about it, but github claims it's 'pending' so perhaps it got stuck somehow... ]

trociny commented:
> Line 69 in doc/radosgw/frontends.rst says column; I suspect you want to say colon.

Thanks. Updated.

cbodley commented on doc/radosgw/frontends.rst:
```diff
@@ -64,6 +64,38 @@ Options
+              ``single_dh_use`` Always create a new key when using tmp_dh parameters.
+
+:Type: String
+:Default: None
```
going back to https://github.com/ceph/ceph/pull/41384, can this default to `no_sslv2:no_sslv3:no_tlsv1:no_tlsv1_1`?

trociny commented:
@cbodley Thanks. Updated. 

I pushed changing the default as a separate commit, because I have plans to backport this PR and I am not sure we will want to change the default behavior during minor releases, and we may want to skip the second commit there.

trociny commented:
@cbodley ping

cbodley reviewed (commented):
can you please add a line describing this change to the `PendingReleaseNotes` file? i wouldn't object to backports as long as it's communicated clearly here

trociny commented on src/rgw/rgw_asio_frontend.cc:
```diff
@@ -801,6 +803,58 @@ int AsioFrontend::init_ssl()
+  }
+
+  if (options) {
+    vector<string> option_list;
+    boost::split(option_list, *options, boost::is_any_of(":"));
```
Ah. I was sure I had deleted them. Thanks!

cbodley reviewed (approved):
nice work, thanks!

k0ste commented:
@cbodley can be this reviewed and merged with #41824? Also we have a "bug" in [#50974](https://tracker.ceph.com/issues/50974) (Nautilus feature). Also to start moving from Beast we need [#51017](https://tracker.ceph.com/issues/51017) and have a bug that we still need ceph.conf file to configure rgw [#50249](https://tracker.ceph.com/issues/50249). Hope all of this can be landed on last of the last 14.2.23 (like a 12.2.13, but better!)
Thanks

trociny commented:
Rebased.

I was preparing a downstream (backport) patch based on this and noticed that one changed from the last review had been added to a wrong commit. So I fixed it to ease the backporting. I checked with `git diff $new_branch $old_branch` that the total change was still the same after my manipulations.
